### ISSUE TITLE
[BUG] DashTable and Wildcard / Pattern Matching callbacks not working

### ISSUE THREAD

**Issue by christianwengert** [plotly/dash#1194]
**Describe your context**

I want to use a datatables 'data' attribute in a callback but using the new pattern matching callbacks. 

This results in a seemingly unrelated JS error (parsing problem?)

```
Failed to execute 'addRule' on 'CSSStyleSheet': Failed to parse the rule '#{"action":"test","table":1} .dash-fixed-column tr { height: 30px; }'.
```

-  replace the result of `pip list | grep dash` below
```
dash                      1.11.0      
dash-bootstrap-components 0.9.2       
dash-canvas               0.1.0       
dash-core-components      1.9.1       
dash-cytoscape            0.1.1       
dash-daq                  0.3.3       
dash-extra-components     0.0.1       
dash-html-components      1.0.3       
dash-leaflet              0.0.3       
dash-planner              0.0.1       
dash-renderer             1.4.0       
dash-table                4.6.2    
```
-  if frontend related, tell us your Browser, Version and OS

    - OS: OSX
    - Browser Any

**Describe the bug**

Here is a test setup to reproduce the bug:

```
import dash
import dash_table
import pandas as pd
import dash_html_components as html
from dash.dependencies import MATCH, Output, Input, ALL

df = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/solar.csv')

app = dash.Dash(__name__)

app.layout = html.Div([
    html.Div(id={'table': 1, 'action': 'output'}),
    dash_table.DataTable(
    id={'table': 1, 'action': 'test'},
    columns=[{"name": i, "id": i} for i in df.columns],
    data=df.to_dict('records'),
)])

@app.callback(Output({'table': MATCH, 'action': 'output'}, 'children'),
              [Input({'table': MATCH, 'action': 'test'}, 'active_cell')])
def test(xy):
    return xy

if __name__ == '__main__':
    app.run_server(debug=True)

```

**Expected behavior**

- No Failed to execute 'addRule' on 'CSSStyleSheet' Error
- Working callback

**Screenshots**

![image](https://user-images.githubusercontent.com/12936057/79241125-eaddd700-7e72-11ea-8b04-e6c6bc9f88cd.png)

![image](https://user-images.githubusercontent.com/12936057/79241157-f5986c00-7e72-11ea-8451-20682653b368.png)

**Comment by alexcjohnson** [comment]
Ah yes - thanks @christianwengert !

@Marc-Andre-Rivet the issue here is that the `dict`/`object` IDs - which we keep as objects in the dash layout - need to be turned into strings when we actually create the associated React element. The way I do it now is just to stringify the object - not exactly `JSON.stringify` since I want the keys sorted, but that's what it looks like. And this form is incompatible with CSS unless properly escaped.

The two possible solutions to this are (1) sanitize the stringified ID in some other way, or (2) ensure any component that wants to use the ID in CSS properly escapes it. (1) seems tricky to do without creating potential collisions, either between string and object IDs or between distinct object IDs. But (2) could potentially require many fixes in different component suites, depending on how widely this kind of dynamic CSS generation is used. That said, it's possible today to create lots of string IDs that would also need escaping - the only characters we exclude from string ids are `{` and `.`, but ` `, `:`, and various others have special meaning in CSS.

https://github.com/plotly/dash/blob/3ec55b5a97fdd34ed8384085df884d86730b1a9c/dash/_validate.py#L95

**Comment by alexcjohnson** [comment]
FYI in case it comes up elsewhere (eg one component trying to reference another one by ID) here's the actual function the renderer uses to stringify IDs:

https://github.com/plotly/dash/blob/8a03f3a1feed780f1ee8d6d47d2d27a00344e5d3/dash-renderer/src/actions/dependencies.js#L114-L126

**Comment by Marc-Andre-Rivet** [comment]
@alexcjohnson If we were to sanitize the id further (eg. `/[~\!@\$%\^&\*\(\)\+=,\.\/';:\"\?><\\\[\]\{\}|`#\s]/` changed to `-`) would this have a negative impact on, say, DevTools display? Is it using the stringified version of the id? Or would we need to (a) stringify/sanitize for some cases and (b) simply stringify for other cases?

The regex above covers a subset of the more frequently used characters that are not supported. We could sanitize further to try and make the id fully compatible with CSS without the need for escaping in user land, but I think it might be better to be stricter on the ids we allow in the first place (eg. HTML4 allows only `[0-9a-zA-Z\._-]` with certain extra limitations.

**Comment by alexcjohnson** [comment]
I don't think we can further restrict the characters in IDs at this point without calling it a breaking change - I was already a bit uneasy about adding `{` to the exclusion, but I did it because the alternatives seemed like they had even worse side-effects.

What do you mean by "escaping in user land"? Isn't this something we can do automatically? Users generally wouldn't be making their own CSS for dict IDs, they would use classes. And if they want to make their own CSS for string IDs we can leave it up to them to either use compatible characters or escape the ID themselves. Seems to me it's only our own internal CSS that's at issue, but it doesn't seem [too hard to fix](https://mathiasbynens.be/notes/css-escapes). The offending rule in @christianwengert 's example I think would just be:
```
#\{\"action\"\:\"test\"\,\"table\"\:1\} .dash-fixed-column tr { height: 30px; }
```

**Comment by Marc-Andre-Rivet** [comment]
Fair enough. In any case, to my knowledge only the table explicitly injects stylesheets into the DOM. Can go for the minimal fix: modify the table so that it escapes the known problematic characters when injecting styles and, as you say above, expect app developers to escapes chars themselves if they want to use special characters, etc.

**Comment by Marc-Andre-Rivet** [comment]
Only the DataTable injects Stylesheet directly, using the `id`. Moving the issue over to the dash-table repo.
